cmd/srcd/cmd/sql_test.go

```diff
@@ -5,24 +5,225 @@ package cmd
+	"regexp"
 	"strings"
 	"testing"
 	"time"
 
+	"github.com/kr/pty"
 	cmdtest "github.com/src-d/engine/cmd/test-utils"
-
+	"github.com/src-d/engine/components"
+	"github.com/src-d/engine/docker"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
 )
 
+var showTablesOutput = sqlOutput(`+--------------+
+| Table        |
++--------------+
+| blobs        |
+| commit_blobs |
+| commit_files |
+| commit_trees |
+| commits      |
+| files        |
+| ref_commits  |
+| refs         |
+| remotes      |
+| repositories |
+| tree_entries |
++--------------+
+`)
+
+var showRepoTableDescOutput = sqlOutput(`+---------------+------+
+| name          | type |
++---------------+------+
+| repository_id | TEXT |
++---------------+------+
+`)
+
+type SQLREPLTestSuite struct {
+	cmdtest.IntegrationSuite
+	testDir string
+}
+
+func TestSQLREPLTestSuite(t *testing.T) {
+	s := SQLREPLTestSuite{}
+	suite.Run(t, &s)
+}
+
+func (s *SQLREPLTestSuite) SetupTest() {
+	var err error
+	s.testDir, err = ioutil.TempDir("", "sql-repl-test")
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (s *SQLREPLTestSuite) TearDownTest() {
+	s.RunCommand(context.Background(), "prune")
+	os.RemoveAll(s.testDir)
+}
+
+func (s *SQLREPLTestSuite) TestREPL() {
+	// When it is not in a terminal, the command reads stdin and exits.
+	// You can see it running:
+	//     $ echo "show tables;" | ./srcd sql
+	// So this test does not really interacts with the REPL prompt, but we still
+	// test that the code that processes each read line is working as expected.
+
+	require := s.Require()
+
+	var out, in bytes.Buffer
+
+	command := s.CommandContext(context.TODO(), "sql")
+	command.Stdout = &out
+	command.Stderr = &out
+	command.Stdin = &in
+
+	io.WriteString(&in, "show tables;\n")
+	io.WriteString(&in, "describe table repositories;\n")
+
+	err := command.Start()
+	require.NoError(err)
+
+	err = command.Wait()
+	require.NoError(err)
+
+	require.Contains(out.String(), showTablesOutput)
+	require.Contains(out.String(), showRepoTableDescOutput)
+}
+
+func (s *SQLREPLTestSuite) TestInteractiveREPL() {
+	require := s.Require()
+
+	command, stdin, ch := s.runInteractiveRepl()
+
+	res := s.runInteractiveQuery(stdin, "show tables;\n", ch)
+	require.Contains(res, showTablesOutput)
+
+	res = s.runInteractiveQuery(stdin, "describe table repositories;\n", ch)
+	require.Contains(res, showRepoTableDescOutput)
+
+	require.NoError(s.exitInteractiveAndWait(10*time.Second, stdin, ch))
+	require.NoError(s.waitMysqlCliContainerStopped(10 * time.Second))
+
+	command.Wait()
+}
+
+func (s *SQLREPLTestSuite) runInteractiveRepl() (*exec.Cmd, *os.File, <-chan string) {
+	s.T().Helper()
+
+	command := s.CommandContext(context.TODO(), "sql")
+
+	ch := make(chan string)
+	cr := cmdtest.NewChannelWriter(ch)
+
+	command.Stdout = cr
+	command.Stderr = cr
+
+	f, err := pty.Start(command)
+	if err != nil {
+		panic(err)
+	}
+
+	mysqlCliReady := make(chan struct{})
```

Comment:
why do you need a channel and goroutine here if you block right after you start the goroutine?